test/controllers/auth/AuthActionSpec.scala

```diff
@@ -0,0 +1,138 @@
+      }
+    }
+
+    "the user has a confidence level too low" must {
+      "redirect the user to the unauthorised page" in {
+
+        val mockAuthConnector = mock[GmpAuthConnector]
+
+        when(mockAuthConnector.authorise(any(),any())(any(), any()))
+          .thenReturn(Future.failed(new InsufficientConfidenceLevel))
+
+        val authAction = new AuthActionImpl(mockAuthConnector,app.configuration)
+        val controller = new Harness(authAction)
+
+        val result = controller.onPageLoad()(FakeRequest("", ""))
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result).get must be("/guaranteed-minimum-pension/unauthorised")
+      }
+    }
+
+    "the user has no psa/psp account " must {
+      "redirect the user to the unauthorised page" in {
+        val mockAuthConnector = mock[GmpAuthConnector]
+
+        when(mockAuthConnector.authorise[Enrolments](any(), any())(any(), any()))
+          .thenReturn(Future.failed(new InsufficientEnrolments))
+
+        val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+        val controller = new Harness(authAction)
+
+        val result = controller.onPageLoad()(FakeRequest("", ""))
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result) must be(Some("/guaranteed-minimum-pension/unauthorised"))
+
+      }
+    }
+
+    "the user is authorised with psa " must {
+      "create an authenticated link for psa" in {
+        val mockAuthConnector = mock[GmpAuthConnector]
+
+        val retrievalResult: Future[Enrolments] =
+          Future.successful(Enrolments(Set(Enrolment("HMRC-PSA-ORG", Seq(EnrolmentIdentifier("PSAID", "someID")), ""))))
+
+        when(mockAuthConnector.authorise[Enrolments](any(), any())(any(), any()))
+          .thenReturn(retrievalResult)
+
+        val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+        val controller = new Harness(authAction)
+
+        val result = controller.onPageLoad()(FakeRequest("", ""))
+        status(result) mustBe OK
+        whenReady(result) {
+          _.header.headers("link") mustBe "someID"
+        }
+      }
+    }
+      "the user is authorised with psp " must {
+        "create an authenticated link for psp" in {
+          val mockAuthConnector = mock[GmpAuthConnector]
+
+          val retrievalResult: Future[Enrolments] =
+            Future.successful(Enrolments(Set(Enrolment("HMRC-PP-ORG", Seq(EnrolmentIdentifier("PPID", "someID")), ""))))
+
+          when(mockAuthConnector.authorise[Enrolments](any(), any())(any(), any()))
+            .thenReturn(retrievalResult)
+
+          val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+          val controller = new Harness(authAction)
+
+          val result = controller.onPageLoad()(FakeRequest("", ""))
+          status(result) mustBe OK
+          whenReady(result) {
+            _.header.headers("link") mustBe "someID"
+          }
+        }
+      }
+
```

Comment:
Need a test for both PSA and PP, where it uses the PSA link by preference

Reply:
Done